### PR TITLE
AP_UAVCAN: add option to send GPS

### PR DESCRIPTION
**Pull request by magicrub**
Add a UAVCAN Options bit to send GNSS packets to that bus. This works with or without a GPS attached but is more interesting when one is :)

### CODE REVIEW

**Comment by magicrub** [comment]
This is on top of https://github.com/ArduPilot/ardupilot/pull/22985 which refactors the uavcan so it doesn't hammer the stack usage so much

**Comment by magicrub** [comment]
Possible Improvements:
- Its a little sloppy to do this packet generation in AP_UAVCAN  It might be better to give AP_GPS_UAVCAN some static functions to populate the packets.
- Add more Option bits to send minimal or all GPS packets
- Use ahrs data instead AP_GPS data

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp [hunk 1]
```diff
@@ -828,6 +867,162 @@ void AP_UAVCAN::notify_state_send()
...
+
+    if (_gnss.last_gps_lib_change_ms != gps.last_message_time_ms()) {
+        _gnss.last_gps_lib_change_ms = gps.last_message_time_ms();
+
+        // GPS library just updated, lets send these now
+        _gnss.last_fix2_ms = 0;
+    }
+
+    // send slow, at 5Hz. If there's an update in the GPS library it will trigger faster, typically 10Hz
+    if (!_gnss.last_fix2_ms || (now_ms - _gnss.last_fix2_ms >= 200)) {
```
Comment (tridge) [inline libraries/AP_UAVCAN/AP_UAVCAN.cpp:888]: i don't think the 2nd part of this (which seems to try to force 5hz) is right, it should send when we really get a new GPS msg

File: libraries/AP_UAVCAN/AP_UAVCAN.h [hunk 2]
```diff
@@ -205,6 +205,7 @@ class AP_UAVCAN : public AP_CANDriver, public AP_ESC_Telem_Backend {
         CANFD_ENABLED             = (1U<<2),
         DNA_IGNORE_UNHEALTHY_NODE = (1U<<3),
         USE_ACTUATOR_PWM          = (1U<<4),
+        SEND_GNSS                 = (1U<<5),
```
Comment (Hwurzburg) [inline libraries/AP_UAVCAN/AP_UAVCAN.h:208]: Need to add this option in the CAN_Dx_US_OPTIONS metadata

**Review by Hwurzburg — DISMISSED** [review]

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp [hunk 3]
```diff
@@ -828,6 +867,162 @@ void AP_UAVCAN::notify_state_send()
     _last_notify_state_ms = AP_HAL::native_millis();
 }
 
+void AP_UAVCAN::gnss_send()
```
Comment (tridge) [inline libraries/AP_UAVCAN/AP_UAVCAN.cpp:870]: maybe break into separate fix2 and yaw send fns

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by tridge** [comment]
do we want to try to avoid sending out on the same CAN bus that GPS came in on?

File: libraries/AP_UAVCAN/AP_UAVCAN.h (hunk 2 above)
Comment (magicrub) [reply libraries/AP_UAVCAN/AP_UAVCAN.h:208]: It already was. What am I missing?

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp (hunk 3 above)
Comment (magicrub) [reply libraries/AP_UAVCAN/AP_UAVCAN.cpp]: done

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp (hunk 1 above)
Comment (magicrub) [reply libraries/AP_UAVCAN/AP_UAVCAN.cpp]: fixed

**Comment by magicrub** [comment]
> do we want to try to avoid sending out on the same CAN bus that GPS came in on?

I'm open to suggestions... maybe next PR?

**Comment by magicrub** [comment]
> > do we want to try to avoid sending out on the same CAN bus that GPS came in on?
> 
> I'm open to suggestions... maybe next PR?

I think I figured out a 1-line solution for it so I added it in

**Comment by magicrub** [comment]
@bugobliterator I've removed the init refactor

**Comment by magicrub** [comment]
Tested on hardware:
- CubeOrange "A" has a real GPS attached and CAN_D1_UC_OPTION = 32
- CubeOrange "B" is connected to "A" via CAN 1 on both ends.
- B GPS data looks good and appears to match A's data.

The output data now looks like like an AP_Peirph... Except during testing I found that Periph is incorrectly sending GNSS:Heading too fast. PR to fix that: https://github.com/ArduPilot/ardupilot/pull/23026

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp [hunk 4]
```diff
@@ -830,6 +862,149 @@ void AP_UAVCAN::notify_state_send()
     _last_notify_state_ms = AP_HAL::native_millis();
 }
 
+void AP_UAVCAN::gnss_send_fix()
+{
+    const AP_GPS &gps = AP::gps();
+
+    if (_gnss.last_gps_lib_fix_ms == gps.last_message_time_ms()) {
```
Comment (tridge) [inline libraries/AP_UAVCAN/AP_UAVCAN.cpp:869]: should do "const uint32_t last_ms = gps.last_message_time_ms()" so we don't ask for it twice

**Review by tridge — COMMENTED** [review]
we need a AP_DRONECAN_SEND_GPS define, and enable it only for 2M flash boards, this costs us 2.8k on Pixhawk1-1M, which is a lot of the remaining flash

**Comment by tridge** [comment]
@bugobliterator shall we delay this for your canard re-work?

**Comment by tridge** [comment]
it probably makes sense for this to go in after sids can re-work #22563

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp (hunk 4 above)
Comment (magicrub) [reply libraries/AP_UAVCAN/AP_UAVCAN.cpp]: fixed

**Comment by magicrub** [comment]
Would appreciate this going in for 4.4. I've talked to @bugobliterator and I'm happy to work with him on regarding his droneCAN rework

**Comment by bugobliterator** [comment]
> Would appreciate this going in for 4.4. I've talked to @bugobliterator and I'm happy to work with him on regarding his droneCAN rework

Yeah, I don't mind going this in, I can work with @magicrub to get this rebasef.

**Review by tridge — APPROVED** [review]

**Comment by rmackay9** [comment]
I've removed the "For-4.5" release because it did make it into the 4.4 releases

**Pull request merged**